### ISSUE TITLE
Knockout api support

### ISSUE THREAD

**Issue by keremkusmezer** [nikhilk/scriptsharp#39]
Is there a possiblity that you will implement also bindings and support for knockout.js?

**Comment by robitar** [comment]
I'm using Knockout with Script# right now, I mentioned in another issue here that you can represent an Observable with a generic delegate, it gives you code which is very close to the native script usage.

That said, I'm currently writing a small library that will 'implement' abstract properties at (script) runtime with knockout observables, allowing you to just reference them as normal properties. Its hacky, but looks good.

**Comment by keremkusmezer** [comment]
Do you have the required c# bindings available? If so can you post it or share on github?

**Comment by robitar** [comment]
Heres what I'm currently using

```
using System;
using System.Collections.Generic;
using System.Html;
using System.Runtime.CompilerServices;

namespace Knockout {
    /// <summary>
    /// Gets or sets an observable value.
    /// </summary>
    /// <typeparam name="T">The type of the underlying value.</typeparam>
    /// <param name="value">The value to set.</param>
    [Imported, IgnoreNamespace]
    public delegate T Observable<T>(params T[] value);

    /// <summary>
    /// Provides access to Knockout JS.
    /// </summary>
    [Imported, IgnoreNamespace, ScriptName("ko")]
    public static class Ko {
        /// <summary>
        /// Creates an observable with an initial value.
        /// </summary>
        /// <typeparam name="T">The type of the observable.</typeparam>
        /// <param name="initialValue">The initial value.</param>
        public static Observable<T> Observable<T>(T initialValue) {
            return null;
        }

        /// <summary>
        /// Creates an observable with no initial value.
        /// </summary>
        /// <typeparam name="T">The type of the observable.</typeparam>
        public static Observable<T> Observable<T>() {
            return null;
        }

        /// <summary>
        /// Maps the values from the source object to the target model, setting or creating observable fields.
        /// </summary>
        /// <typeparam name="T">The model type.</typeparam>
        /// <param name="source">A data object whose properties will be mapped over.</param>
        /// <param name="mapping">Options for fine tuning the mapping.</param>
        /// <param name="model">The model instance to apply the values to.</param>
        [ScriptAlias("ko.mapping.fromJS")]
        public static void Map<T>(object source, object mapping, T model) {
        }

        /// <summary>
        /// Updates mapped values after a previous call to <see cref="Map{T}(object, object, T)"/>.
        /// </summary>
        /// <typeparam name="T">The model type.</typeparam>
        /// <param name="model">The model instance to apply the values to.</param>
        /// <param name="source">A data object whose properties will be mapped over.</param>
        [ScriptAlias("ko.mapping.updateFromJS")]
        public static void UpdateMap<T>(T model, object source) {
        }

        /// <summary>
        /// Sets up bindings for the model, scoped to the given root element.
        /// </summary>
        /// <param name="model">The model.</param>
        /// <param name="rootElement">The element to bind to.</param>
        public static void ApplyBindings(object model, Element rootElement) {
        }

        /// <summary>
        /// Sets up bindings for the model.
        /// </summary>
        /// <param name="model">The model.</param>
        public static void ApplyBindings(object model) {
        }

        /// <summary>
        /// Creates an empty observable array.
        /// </summary>
        /// <typeparam name="T">The type of items in the array.</typeparam>
        public static ObservableArray<T> ObservableArray<T>() {
            return null;
        }

        /// <summary>
        /// Creates an observable array with some initial items.
        /// </summary>
        /// <typeparam name="T">The type of items in the array.</typeparam>
        /// <param name="initialItems">A sequence of initial items.</param>
        public static ObservableArray<T> ObservableArray<T>(IEnumerable<T> initialItems) {
            return null;
        }

        /// <summary>
        /// Returns true if the value is an Observable, false otherwise.
        /// </summary>
        /// <param name="value">The value to check.</param>
        public static bool IsObservable(object value) {
            return false;
        }

        /// <summary>
        /// If the provided value is an observable, return its value, otherwise just pass it through.
        /// </summary>
        /// <param name="value">The potential observalbe.</param>
        [ScriptAlias("ko.utils.unwrapObservable")]
        public static object UnwrapObservable(object value) {
            return null;
        }

        /// <summary>
        /// Creates an observable with a value computed from one or more other observables.
        /// </summary>
        /// <typeparam name="T">The type of the observable value/</typeparam>
        /// <param name="function">A function to compute the value.</param>
        public static Observable<T> DependentObservable<T>(Func<T> function) {
            return null;
        }

        /// <summary>
        /// Raises an event on the given element.
        /// </summary>
        /// <param name="element">An element.</param>
        /// <param name="type">The type, or name, of the event, such as 'change'.</param>
        [ScriptAlias("ko.utils.triggerEvent")]
        public static void TriggerEvent(Element element, string type) {
        }
    }

    /// <summary>
    /// Detect and respond to changes of a collection of items.
    /// </summary>
    /// <typeparam name="T">Type of the items to manage.</typeparam>
    [Imported, IgnoreNamespace]
    public abstract class ObservableArray<T> {
        /// <summary>
        /// Adds the value and notifies observers.
        /// </summary>
        /// <param name="value">The value to add.</param>
        public abstract void Push(T value);

        /// <summary>
        /// Returns the index of the first array item that equals the value, or -1 if there is no match.
        /// </summary>
        /// <param name="value">The value to check.</param>
        public abstract int IndexOf(T value);

        /// <summary>
        /// Returns elements from start index to end index.
        /// </summary>
        /// <param name="start">Starting point of the sequence, if negative then it starts from the end.</param>
        /// <param name="end">End point of the sequence.</param>
        public abstract T[] Slice(int start, int end);

        /// <summary>
        /// Returns elements from start index to the end of the array.
        /// </summary>
        /// <param name="start">Starting point of the sequence, if negative then it starts from the end.</param>
        public abstract T[] Slice(int start);

        /// <summary>
        /// Removes the last value from the array and returns it.
        /// </summary>
        public abstract T Pop();

        /// <summary>
        /// Removes the first value from the array and returns it
        /// </summary>
        public abstract T Shift();

        /// <summary>
        /// Inserts the value at the beginning of the array.
        /// </summary>
        /// <param name="value">The value to insert.</param>
        public abstract void Unshift(T value);

        /// <summary>
        /// Reverses the order of the array.
        /// </summary>
        public abstract void Reverse();

        /// <summary>
        /// Performs a default alphanumeric sort on the elements of the array.
        /// </summary>
        public abstract void Sort();

        /// <summary>
        /// Performs a sort using the comparator function.
        /// </summary>
        /// <param name="comparator">The comparing function.</param>
        public abstract void Sort(Func<T, T, int> comparator);

        /// <summary>
        /// Removes all values that match the given parameter and returns them.
        /// </summary>
        /// <param name="value">The value to remove.</param>
        public abstract T[] Remove(T value);

        /// <summary>
        /// Removes all values that satisfy the predicate and returns them.
        /// </summary>
        /// <param name="predicate">The removal predicate.</param>
        public abstract T[] Remove(Func<T, bool> predicate);

        /// <summary>
        /// Removes all values that satisfy the given parameters and returns them.
        /// </summary>
        /// <param name="values">An array of items to remove.</param>
        public abstract T[] RemoveAll(params T[] values);

        /// <summary>
        /// Marks all values that match the given parameter as deleted.
        /// </summary>
        /// <param name="value">The value to mark as deleted.</param>
        public abstract void Destroy(T value);

        /// <summary>
        /// Marks all values that satisfy the predicate as deleted.
        /// </summary>
        /// <param name="predicate">The predicate.</param>
        public abstract void Destroy(Func<T, bool> predicate);

        /// <summary>
        /// Marks all values that satisfy the given parameters as deleted.
        /// </summary>
        /// <param name="values">An array of items to destroy.</param>
        public abstract void DestroyAll(params T[] values);

        public static implicit operator Observable<T[]>(ObservableArray<T> array) {
            return null;
        }
    }
}
```

**Comment by keremkusmezer** [comment]
Thx a lot for the implementation and mapping sharing, maybe we should share that with nikhilk also?

**Comment by nikhilk** [comment]
I did see this -
I will consider including support ... (need to think through the observable stuff I've been planning a bit).

One other idea to consider ... and see if it works:

```
[Imported, IgnoreNamespace]
public sealed class Observable<T> {
    private Observable() { }

    [ScriptName("")]
    public T GetValue() { return default(T); }

    [ScriptName("")]
    public void SetValue(T value) { }
}
```

instead of a delegate with a params array argument that isn't as concrete on the intended signature.

If you do try it, I'd love to hear if it works.

**Comment by robitar** [comment]
Just tried it now, and I'm surprised, but it works. I would have thought that the expression would end up as:

c#

``` c#
field.SetValue(5)
```

expected

``` js
field.(5)
```

But it seems smart enough to leave the dot operator out too:

actual

``` js
field(5)
```

I was sure I tried something like this but couldn't get around that issue, odd.

I definitely prefer this approach, now that it seems to work. Even though observables are functions in script, they have methods, and there's an inheritance relationship with observable arrays etc.

**Comment by nikhilk** [comment]
I will go ahead and include an import library for knockout, since it seems in high demand, and while what I am thinking of is still in early stages... and even when it is ready will be a choice/option.

Is there an API listing style doc for knockout. I only saw conceptual docs. Is the API above complete?

**Comment by sagacity** [comment]
There is some documentation on http://knockoutjs.com/documentation/introduction.html but the full API can be extracted from the main `knockout.js` and `knockout.mapping.js` files by scanning for the `ko.exportSymbol` and `ko.exportProperty` directives.

**Comment by nikhilk** [comment]
Yes, I've seen the conceptual doc. I was hoping there was something akin to jQuery API docs. Its somewhat suboptimal to piece out the API from scanning the actual script, list all the export calls, look each one up all possible parameter/return value signatures.

The question about completeness of above API still remains, and if so it is already a better API reference. Thats the nice thing about c# code... :-)

**Comment by robitar** [comment]
This is something I found quite frustrating, no API reference anywhere for Knockout.

The bindings I wrote above were made up as I went through the conceptual documentation on the site, on an as-needed basis. I checked the source a few times but it's in now way complete, hardly any of the utility functions are in there. I could go through the rest and fill it out, to save you the hassle of doing it ;)

**Comment by nikhilk** [comment]
I wrote an initial version of the API based on what you had above and based on what I saw.

**Comment by robitar** [comment]
Just noticed after I posted. As I'm planning to make heavy use of Knockout, I'll switch over from my hand rolled bindings to this one and expand it as I run into corner cases.

**Comment by nikhilk** [comment]
Fix included in 0.7.2.0 release.

**Comment by andrewharry** [comment]
I've added a pull request for a updated and more complete binding for knockout 1.3rc
